rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+different proofs, even where the 'position' of the same side-loaded table in
+the combined table may vary between proofs. For example, we could consider a
+'folding' circuit `C` which takes 2 side-loaded tables `t_in` and `t_out`,
+where we prove
+```
+C(t_0, t_1)
+C(t_1, t_2)
+C(t_2, t_3)
+...
+```
+and the 'result' of the fold is the final `t_n`. In each pair of executions in
+the sequence, we find that `t_i` is used as both `t_in` and `t_out`, but we
+would like to expose the same table as the value for each.
+
+To achieve this, we use a permutation argument to 'place' the values from each
+side-loaded table at the correct position in the final table. Recall that every
+value
+```
+LookupTable - FixedAndRuntimeTable
+```
+is either 0 (where no side-loaded table contributes a value) or the value of
+some row in `SideLoadedCombined(i) * joint_combiner` from the table `i`.
+
+For the `0` differences, we calculate a permutation between them all and the
+single constant `0`, to ensure that `LookupTable = FixedAndRuntimeTable`. For
+all other values, we set-up a permutation based upon the side-loaded table
+relevant to each value.
+
+Thus, we build the permutation accumulator
+```
+LookupPermutation
+ * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+     + gamma + beta * Sigma(7))
+ * (SideLoadedCombined(0) + gamma + beta * Sigma(8))
+ * (SideLoadedCombined(1) + gamma + beta * Sigma(9))
+ * (SideLoadedCombined(2) + gamma + beta * Sigma(10))
+ * (SideLoadedCombined(3) + gamma + beta * Sigma(11))
+ =
+   LookupPermutation[prev]
+   * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+       + gamma + x * beta * shift[7])
+   * (SideLoadedCombined(0) + gamma + x * beta * shift[8])
+   * (SideLoadedCombined(1) + gamma + x * beta * shift[9])
+   * (SideLoadedCombined(2) + gamma + x * beta * shift[10])
+   * (SideLoadedCombined(3) + gamma + x * beta * shift[11])
+```
+where `shift[0..7]` is the existing `shift` vector used for the witness
+permutation, and the additional values above are chosen so that
+`shift[i] * w^j` are distinct for all `i` and `j`, and `Sigma(_)` are the
+columns representing the permutation. We then assert that
+```
+Lagrange(0)
+  * (LookupPermutation * (0 + gamma + beta * zero_sigma)
+    - (0 + gamma + beta * shift[len(shift)-1]))
+  = 0
+```
+to mix in the contributions for the constant `0` value, and
+```
+Lagrange(n-ZK_ROWS) * (LookupPermutation - 1) = 0
+```
+to ensure that the permuted values cancel.
+
+Note also that the permutation argument allows for interaction **between**
+tables, as well as injection of values into the combined lookup table. As a
+result, the permutation argument can be used to 'copy' parts of one table into
+another, which is likely to be useful in examples like the 'fold' one above, if
+some data is expected to be shared between `t_in` and `t_out`.
+
+### Permutation argument for cheap array lookups
+
+It would be useful to add data to the runtime table using the existing
```

Comment:
I can imagine how to add data using a lookup, but it would be helpful to get an quick high-level intuition how to add data using permutation :-)

Reply:
You could imagine a primitive `store_array(idx, [val_0, val_1, ..., val_n])` in snarky that creates a permutation between the previous position of `val_i` and the `idx+i`th entry of the table, in the same way that we create a permutation between subsequent uses of a value `val` between different rows in the main witness. This would be functionally the same as `store(idx, val_0)`, `store(idx+1, val_1)`, ..., `store(idx+n, val_n)` where `store` uses lookups, but doing it this way lets us avoid using lots of lookups across lots of rows to initialise the storage.

For an explicit example, consider
```
store_array(idx, [{row= 155, col= 2}, {row= 205, col=7}])
```
By considering the values in the runtime table as 'column 12', we can build the equivalence sets as usual, and then we get the permutations
```
[{row= 155, col= 2}, {row= idx, col= 12}, ....]
[{row= 205, col= 7}, {row= idx+1, col= 12}, ...]
```

Naturally this is only really usable when `idx` is fixed by the constraint system, but it will usually be possible to do this by 'allocating some blocks' in the runtime table corresponding to each array that you want to load.